loki/backend/cgen.py

```diff
@@ -174,7 +174,18 @@ def visit_Subroutine(self, o, **kwargs):
                 aptr += ['']
         arguments = [f'{self.visit(a.type, **kwargs)} {p}{a.name.lower()}'
                      for a, p in zip(o.arguments, aptr)]
-        header += [self.format_line('int ', o.name, '(', self.join_items(arguments), ') {')]
+
+        # check whether to return something and define function return type accordingly
+        return_var = None
+        if o.is_function:
+            # Determine function result variable name
+            if not (result_name := o.result_name):
+                result_name = o.name.replace('_c', '')
+            if result_name in o.variable_map:
+                return_var = o.variable_map[result_name]
+        return_type = c_intrinsic_type(return_var.type) if return_var is not None else 'void'
```

Comment:
This reimplements the logic in https://github.com/ecmwf-ifs/loki/blob/31eabbee02531a7e5f25030cdf60b27ff0a7cfbd/loki/subroutine.py#L324-L332
but handles the special case that the function name no longer matches the name of the return variable.  But that seems like an inconsistent IR state.

I would suggest to instead handle that renaming in the f2c transformation before this line:
https://github.com/ecmwf-ifs/loki/blob/31eabbee02531a7e5f25030cdf60b27ff0a7cfbd/loki/transform/fortran_c_transform.py#L413
with something like this:
```py
if kernel.is_function and kernel.result_name is None:
    kernel.result_name = kernel.name
```

I think that should allow you to simply write here then
```suggestion
        if o.is_function:
            return_type = c_intrinsic_type(o.return_type)
        else:
            return_type = 'void'
```